src/coreclr/jit/optimizer.cpp

```diff
@@ -8898,6 +8898,581 @@ bool Compiler::optIsRangeCheckRemovable(GenTree* tree)
+    // Fix 388436 ARM JitStress WP7
+    c1->gtFlags &= ~GTF_SET_FLAGS;
+    c2->gtFlags &= ~GTF_SET_FLAGS;
+
+    // The new top level node that we just created does feed directly into
+    // a comparison against zero, so set the GTF_SET_FLAGS bit so that
+    // we generate an instruction that sets the flags, which allows us
+    // to omit the cmp with zero instruction.
+
+    // Request that the codegen for cmpOp1 sets the condition flags
+    // when it generates the code for cmpOp1.
+    //
+    cmpOp1->gtRequestSetFlags();
+#endif
+
+    flowList* edge1 = fgGetPredForBlock(b1->bbJumpDest, b1);
+    flowList* edge2;
+
+    /* Modify the target of the conditional jump and update bbRefs and bbPreds */
+
+    if (sameTarget)
+    {
+        edge2 = fgGetPredForBlock(b2->bbJumpDest, b2);
+    }
+    else
+    {
+        edge2 = fgGetPredForBlock(b2->bbNext, b2);
+
+        fgRemoveRefPred(b1->bbJumpDest, b1);
+
+        b1->bbJumpDest = b2->bbJumpDest;
+
+        fgAddRefPred(b2->bbJumpDest, b1);
+    }
+
+    noway_assert(edge1 != nullptr);
+    noway_assert(edge2 != nullptr);
+
+    BasicBlock::weight_t edgeSumMin = edge1->edgeWeightMin() + edge2->edgeWeightMin();
+    BasicBlock::weight_t edgeSumMax = edge1->edgeWeightMax() + edge2->edgeWeightMax();
+    if ((edgeSumMax >= edge1->edgeWeightMax()) && (edgeSumMax >= edge2->edgeWeightMax()))
+    {
+        edge1->setEdgeWeights(edgeSumMin, edgeSumMax, b1->bbJumpDest);
+    }
+    else
+    {
+        edge1->setEdgeWeights(BB_ZERO_WEIGHT, BB_MAX_WEIGHT, b1->bbJumpDest);
+    }
+
+    /* Get rid of the second block (which is a BBJ_COND) */
+
+    noway_assert(b1->bbJumpKind == BBJ_COND);
+    noway_assert(b2->bbJumpKind == BBJ_COND);
+    noway_assert(b1->bbJumpDest == b2->bbJumpDest);
+    noway_assert(b1->bbNext == b2);
+    noway_assert(b2->bbNext);
+
+    fgUnlinkBlock(b2);
+    b2->bbFlags |= BBF_REMOVED;
+
+    // If b2 was the last block of a try or handler, update the EH table.
+
+    ehUpdateForDeletedBlock(b2);
+
+    /* Update bbRefs and bbPreds */
+
+    /* Replace pred 'b2' for 'b2->bbNext' with 'b1'
+     * Remove  pred 'b2' for 'b2->bbJumpDest' */
+
+    fgReplacePred(b2->bbNext, b2, b1);
+
+    fgRemoveRefPred(b2->bbJumpDest, b2);
+
+    /* Update the block numbers and try again */
+
+    *change = true;
+    /*
+                do
+                {
+                    b2->bbNum = ++n1;
+                    b2 = b2->bbNext;
+                }
+                while (b2);
+    */
+
+    // Update loop table
+    fgUpdateLoopsAfterCompacting(b1, b2);
+
+#ifdef DEBUG
+    if (verbose)
+    {
+        printf("Folded %sboolean conditions of " FMT_BB " and " FMT_BB " to :\n", c2->OperIsLeaf() ? "" : "non-leaf ",
+               b1->bbNum, b2->bbNum);
+        gtDispStmt(s1);
+        printf("\n");
+    }
+#endif
+}
+
+/*
+    Optimizes boolean when the second BB is BBJ_RETURN
+*/
+void Compiler::optOptimizeBoolsBbjReturn(BasicBlock* b1, BasicBlock* b2, BasicBlock* b3, bool* change)
```

Comment:
This at least superficially looks a lot like the above optOptimizeBoolsBbjCond. E.g., it's odd to copy the whole "#if FEATURE_SET_FLAGS" including reference to the ancient WP7 bug database. Can you refactor them to share anything?

btw, I think you should remove the "Bbj" part of the names, and use names like `optOptimizeBoolsCondBlock` and `optOptimizeBoolsReturnBlock` instead. (And make sure the function header comments make it very clear how they are different, and the same)

Reply:
Refactored the code to share the common code between the Conditional block and Return block boolean optimization.